packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -574,7 +586,16 @@ export class ConnectionManager implements IConnectionManager {
 				if (retryDelayFromError !== undefined) {
 					this.props.reconnectionDelayHandler(retryDelayFromError, origError);
 				}
-				await waitForConnectedState(delayMs);
+
+				// First, wait for the amount of time we think is appropriate.
+				await new Promise<void>((resolve) => {
+					setTimeout(resolve, delayMs);
+				});
+
+				// Then, if we believe we're offline, we assume there's no point in trying until we at least think we're online.
+				// NOTE: This isn't strictly true for drivers that don't require network (e.g. local driver).  Really this logic
+				// should probably live in the driver.
+				await waitForOnline();
```

Comment:
Glancing at the usage, I think the intention of the old code was - if network request failed, then retry with some polling frequency. But in case we are offline (and thus network error was caused by being offline), disregard polling frequency if we detect establishment of connection.

It's probably does not matter that much with our max retry of 8 seconds, but if it was in minutes, then new behavior might result (from end-user POV) as a regression, as we would not reconnect immediately on network coming back to life.

I agree with confusing behavior hidden in one function, but maybe the right fix here to replace waitForConnectedState(minDelay) usage with Promise.race(waitForOnline(), delay(miDelay)) - that's what callers expect, I think.

Reply:
One thing I was trying to avoid is violating the `error.retryAfterSeconds` - seems like regardless of network coming online/offline we should not try again sooner than the error specifies.  Maybe I'll split out the `retryDelayFromError` from the `delayMs` so we can wait the full duration the error requests, but otherwise race like you say.

Reply:
Logically, we should never have retryAfterSeconds & being offline. That, of course, assumes that offline signal is accurate (which is not the case, but I do not if it produces false positives - it definitely produces false negatives).

So while your concern is valid, I'm not sure how much we should care.
But if we want to take care of that combo, we likely need to make more invasive changes to consuming code, and separate dealing with retryAfterSeconds vs. polling frequency (when we do not have retryAfterSeconds).
I'd personally prefer for code to be simpler :)

Reply:
Latest includes this improvement for ConnectionManager (no race required - no matter what every path should wait for online before retry).  Taking a look at ParallelRequests now.